ocaml/xapi-storage-script/lib.ml

```diff
@@ -0,0 +1,259 @@
+                path (Printexc.to_string exn)
+            in
+            Lwt.fail (Failure msg)
+        )
+    in
+    loop [] path
+end
+
+module Signal = struct
+  type t = int
+
+  let to_string s = Fmt.(str "%a" Dump.signal s)
+end
+
+module Process = struct
+  module Output = struct
+    type exit_or_signal = Exit_non_zero of int | Signal of Signal.t
+
+    type t = {
+        exit_status: (unit, exit_or_signal) Result.t
+      ; stdout: string
+      ; stderr: string
+    }
+
+    let exit_or_signal_of_unix = function
+      | Unix.WEXITED 0 ->
+          Ok ()
+      | WEXITED n ->
+          Error (Exit_non_zero n)
+      | WSIGNALED n ->
+          Error (Signal n)
+      | WSTOPPED n ->
+          Error (Signal n)
+  end
+
+  let create ~env ~prog ~args =
+    let args = Array.of_list (prog :: args) in
+    let cmd = (prog, args) in
+
+    let env =
+      Unix.environment ()
+      |> Array.to_seq
+      |> Seq.map (fun kv ->
+             let k, v = Scanf.sscanf kv "%s@=%s" (fun k v -> (k, v)) in
+             (k, v)
+         )
+      |> StringMap.of_seq
+      |> StringMap.add_seq (List.to_seq env)
+      |> StringMap.to_seq
+      |> Seq.map (fun (k, v) -> Printf.sprintf "%s=%s" k v)
+      |> Array.of_seq
+    in
+
+    Lwt_process.open_process_full ~env cmd
+
+  let close chan () = Lwt_io.close chan
+
+  let send chan data =
+    Lwt.finalize (fun () -> Lwt_io.write chan data) (close chan)
+
+  let receive chan = Lwt.finalize (fun () -> Lwt_io.read chan) (close chan)
+
+  let run ~env ~prog ~args ~input =
+    let p = create ~env ~prog ~args in
+    let sender = send p#stdin input in
+    let receiver_out = receive p#stdout in
+    let receiver_err = receive p#stderr in
+    Lwt.catch
+      (fun () ->
+        let receiver = Lwt.both receiver_out receiver_err in
+        Lwt.both sender receiver >>= fun ((), (stdout, stderr)) ->
+        p#status >>= fun status ->
+        let exit_status = Output.exit_or_signal_of_unix status in
+        Lwt.return {Output.exit_status; stdout; stderr}
+      )
+      (function
+        | Lwt.Canceled as exn ->
+            Lwt.cancel receiver_out ; Lwt.cancel receiver_err ; Lwt.fail exn
+        | exn ->
+            Lwt.fail exn
+        )
+end
+
+module DirWatcher = struct
+  type event = Modified of string | Changed
+
+  let create path =
+    Lwt_inotify.create () >>= fun desc ->
+    let watches = Hashtbl.create 32 in
+    let selectors =
+      Inotify.[S_Close; S_Create; S_Delete; S_Delete_self; S_Modify; S_Move]
+    in
+    Lwt_inotify.add_watch desc path selectors >>= fun watch ->
+    (* Deduplicate the watches by removing the previous one from inotify and
+       replacing it in the table *)
+    let maybe_remove =
+      if Hashtbl.mem watches watch then
+        Lwt_inotify.rm_watch desc watch
+      else
+        Lwt.return_unit
+    in
+    maybe_remove >>= fun () ->
+    Hashtbl.replace watches watch path ;
+    Lwt.return (watches, desc)
+
+  let rec read (watches, desc) =
```

Comment:
There is a `rec` but I can't see a recursive call to itself in the body.

Reply:
Probably a remnant of the previous, more complex implementation that needed it